File: 0116_Populating_Next_Right_Pointers_in_Each_Node/solution.java
package cn.huang.leetcode;

import cn.huang.leetcode.common.TreeLinkNode;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

/*
116. Populating Next Right Pointers in Each Node

    You may only use constant extra space.
    Recursive approach is fine, implicit stack space does not moves as extra space for this problem.
    You may assume that it is a perfect binary tree (ie, all leaves are at the same level, and every parent has two children).

 */
public class LC_0116_PopulatingNextRightPointersinEachNode {

    /*
    BFS, easy but not constant space
    Complexity: time O(N) space O(N) - queue
     */
    public void connect(TreeLinkNode root) {
        if (root == null)
            return;

        Queue<TreeLinkNode> q = new LinkedList<>();
        q.add(root);

        while (!q.isEmpty()) {
            int count = q.size();
            List<TreeLinkNode> layer = new ArrayList<>();
            for (int i = 0; i < count; i++) {
                TreeLinkNode node = q.poll();
                layer.add(node);
                if (node.left != null)
                    q.add(node.left);
                if (node.right != null)
                    q.add(node.right);
            }

            for (int i = 0; i < count - 1; i++) {
                layer.get(i).next = layer.get(i + 1);
            }
        }
    }

    /*
    http://rainykat.blogspot.com/2017/02/leetcodef-117-populating-next-right.html
    Do not use layer
     */
    public void connect_better(TreeLinkNode root) {
        if (root == null)
            return;

        Queue<TreeLinkNode> q = new LinkedList<>();
        q.add(root);

        while (!q.isEmpty()) {
            int count = q.size();
            for (int i = 0; i < count; i++) {
                TreeLinkNode node = q.poll();
                if (i < count - 1) {
                    TreeLinkNode next = q.peek();
                    node.next = next;
                }
                if (node.left != null)
                    q.add(node.left);
                if (node.right != null)
                    q.add(node.right);
            }
        }
    }

    public void connect_recursion(TreeLinkNode root) {
        if (root == null)
            return;
        if (root.left != null) {
            root.left.next = root.right;
        }
        // have left sub tree and right sub tree
        if (root.next != null && root.right != null) {
            root.right.next = root.next.left;   // connect left sub tree and right sub tree
        }
        connect_recursion(root.left);
        connect_recursion(root.right);
    }
}

File: 0116_Populating_Next_Right_Pointers_in_Each_Node/solution.py
'''
116. Populating Next Right Pointers in Each Node

Given a binary tree

    struct TreeLinkNode {
      TreeLinkNode *left;
      TreeLinkNode *right;
      TreeLinkNode *next;
    }
Populate each next pointer to point to its next right node. If there is no next right node, the next pointer should be set to NULL.

Initially, all next pointers are set to NULL.

Note:

You may only use constant extra space.
You may assume that it is a perfect binary tree (ie, all leaves are at the same level, and every parent has two children).
For example,
Given the following perfect binary tree,
         1
       /  \
      2    3
     / \  / \
    4  5  6  7
After calling your function, the tree should look like:
         1 -> NULL
       /  \
      2 -> 3 -> NULL
     / \  / \
    4->5->6->7 -> NULL
'''


class Solution(object):
    #  Pass at first try!
    def connect(self, root):
        """
        :type root: TreeLinkNode
        :rtype: nothing
        """

        if root is None:
            return

        stack = [root]
        while len(stack) > 0:
            new_level = []
            for i in xrange(len(stack) - 1):
                stack[i].next = stack[i + 1]
                if stack[i].left:
                    new_level.append(stack[i].left)
                if stack[i].right:
                    new_level.append(stack[i].right)

            stack[-1].next = None
            if stack[-1].left:
                new_level.append(stack[-1].left)
            if stack[-1].right:
                new_level.append(stack[-1].right)
            stack = new_level
